### PR TITLE
ARROW-6678: [C++][Parquet] Binary data stored in Parquet metadata must be base64-encoded to be UTF-8 compliant

### PR DESCRIPTION
**Pull request by wesm**
I have added a simple base64 implementation (Zlib license) to arrow/vendored from

https://github.com/ReneNyffenegger/cpp-base64

### CODE REVIEW

**Comment by kou** [comment]
It seems that Boost provides base64 related iterators:

  * https://github.com/boostorg/serialization/blob/develop/include/boost/archive/iterators/base64_from_binary.hpp
  * https://github.com/boostorg/serialization/blob/develop/include/boost/archive/iterators/binary_from_base64.hpp

Can we use them? Or should we bundle new base64 library?

FYI: It seems that Boost bundles the same bse64 library since 1.66: https://github.com/boostorg/beast/blob/develop/include/boost/beast/core/detail/base64.ipp

**Comment by wesm** [comment]
I think the strategy here is a stopgap so that we can release. I think we are trying to depend less in general on Boost so either way vendoring a base64 implementation may be a good idea

**Comment by kou** [comment]
I see.

**Comment by wesm** [comment]
This needs exports for MSVC. Adding them

**Comment by wesm** [comment]
Here's running builds

Appveyor: https://ci.appveyor.com/project/wesm/arrow/builds/27646405
Travis CI: https://travis-ci.org/wesm/arrow/builds/589252353

Can be merged after a bit more of the builds run if deemed acceptable

File: cpp/src/arrow/vendored/base64.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+   Permission is granted to anyone to use this software for any purpose,
+   including commercial applications, and to alter it and redistribute it
+   freely, subject to the following restrictions:
+
+   1. The origin of this source code must not be misrepresented; you must not
+      claim that you wrote the original source code. If you use this source code
+      in a product, an acknowledgment in the product documentation would be
+      appreciated but is not required.
+
+   2. Altered source versions must be plainly marked as such, and must not be
```
Comment (emkornfield) [inline cpp/src/arrow/vendored/base64.cpp]: I assume the source code hasn't been modified?

Comment (wesm) [reply cpp/src/arrow/vendored/base64.cpp]: Unmodified

**Comment by emkornfield** [comment]
All C++ related builds have passed.  I think this can merged.

As follow-ups it would be nice to have:
1.  Tests for forward compatibility if the encoding changes in some way
2.  Possibly adding integration smoke  tests (including the one used that found this issue to some place in our CI).  Maybe just running the parquet-mr tool?

File: cpp/src/arrow/vendored/base64.cpp (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/vendored/base64.cpp]: To be pedantic, the only difference is adding the `arrow::util` namespace.

File: cpp/src/arrow/vendored/base64.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  size_t in_len = encoded_string.size();
+  int i = 0;
+  int j = 0;
+  int in_ = 0;
+  unsigned char char_array_4[4], char_array_3[3];
+  std::string ret;
+
+  while (in_len-- && ( encoded_string[in_] != '=') && is_base64(encoded_string[in_])) {
+    char_array_4[i++] = encoded_string[in_]; in_++;
+    if (i ==4) {
```
Comment (emkornfield) [inline cpp/src/arrow/vendored/base64.cpp]: nit space is off here.

**Comment by wesm** [comment]
Whether UTF-8 is validated may depend on the Thrift library. It was only caught incidentally in a Python-based Thrift decoder because Python did `decode('utf-8')` on the bytes and failed

File: cpp/src/arrow/vendored/base64.cpp (hunk 2 above)
Comment (wesm) [reply cpp/src/arrow/vendored/base64.cpp]: This is the case in the source file https://github.com/ReneNyffenegger/cpp-base64/blob/master/base64.cpp#L97

Comment (wesm) [reply cpp/src/arrow/vendored/base64.cpp]: (we don't clang-format the files in vendored/*)

**Comment by emkornfield** [comment]
+1, thanks.

**Pull request closed without merging**